### PR TITLE
Vvmax

### PR DESCRIPTION
**Pull request by cwjames1983**
This is for Wayne's V/Vmax paper: see papers/Vvmax

It uses FRB redshifts from the ICS paper.

I've pulled in main - there were, amazingly, no conflicts to resolve!

"zdm" is only used for calculating, well, z(DM), and also luminosity distances, but it was a very convenient library for doing so.

### CODE REVIEW

File: papers/Vvmax/Data/all_unloc_frb_data.dat [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# FRB                 DM        Freq          S/N    S/Nth   Dmg    dt         w         theta_d     Fnu       Zmac Nants RF
```
Comment (profxj) [inline papers/Vvmax/Data/all_unloc_frb_data.dat:1]: Is this data that others will want to use for future analyses?

if so, let's "expose" it by having it in `zdm/data/`

File: papers/Vvmax/RLFs.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
+import numpy as np
```
Comment (profxj) [inline papers/Vvmax/RLFs.py:1]: similar to the data, if any of the code that follows will be useful to the community, it should be in the main repo folder, e.g.

`zdm/vvmax/RLFs.py`

File: papers/Vvmax/RLFs.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+# normalisation energy - no physical meaning, just makes units more sensible
+E0=1e22
+
+
+def make_first_guess(E,L, params,method):
+    """
+    Makes a first guess by setting the amplitude
+    to that of the first bin
+    
+    Params are always gamma,Emax
```
Comment (profxj) [inline papers/Vvmax/RLFs.py:14]: define `E, L, method`
  include data type if you can

File: papers/Vvmax/RLFs.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        mag = power_law(E[0],1.,params[0],params[1])
+        guess = [L[0]/mag,params[0],params[1]]
+    
+    
+    return guess
+    
+    
+
+
+def logIntegrateSchechter(log10E,A,gamma,log10Emax):
```
Comment (profxj) [inline papers/Vvmax/RLFs.py:36]: I'm not going to nit pick the doc strings unless these wind up 
in the main Repo folder (i.e. `zdm`).  

So if you move them there (it seems you probably should), then 
add short descriptions of each input paraemter

File: papers/Vvmax/calc_vvmax.py [hunk 5]
```diff
@@ -0,0 +1,764 @@
+"""
+Script to calculate V and Vmax for CRAFT FRBs
```
Comment (profxj) [inline papers/Vvmax/calc_vvmax.py:2]: is the code generic enough to be ueseful for other experiments?

if so, all the more reason to have this exposed in the main repo

File: papers/Vvmax/calc_vvmax.py [hunk 6]
```diff
@@ -0,0 +1,764 @@
...
+
+# initialises cosmology and distance measures 
+
+
+def main(fname,outfile,Nsfr,alpha=0.,loc=False,minz=None,write=True,
+    laststrings=None,LimZ=None,zfrac = 0.):
+    """
+    
+    """
+    global macquartz
```
Comment (profxj) [inline papers/Vvmax/calc_vvmax.py:31]: `global`s are so...  lazy  ;)

File: papers/Vvmax/calc_vvmax.py [hunk 7]
```diff
@@ -0,0 +1,764 @@
...
+    # i.e. it ignores the fact the intrinsic width behaves differently...
+    # eff0 = (1+z0)**-0.5
+    # effz = (1+ztrials)**-0.5
+    
+    sz = s0 * (effz/eff0) * modFz
+    
+    return sz,eff0,effz,modFz
+    
+
+def calc_zmax(s0,z0,DM0,DMcosmic0,w0,freq,tsamp,state,Mz,MDM,alpha):
```
Comment (profxj) [inline papers/Vvmax/calc_vvmax.py:330]: this looks like a very useful method for the community
   there is a somewhat similar one in the `FRB` repo already

File: papers/Vvmax/calc_vvmax.py [hunk 8]
```diff
@@ -0,0 +1,764 @@
...
+    plt.scatter(DM, Fluences/Fnu)
+    plt.xlabel("DM")
+    plt.ylabel("James F / Arcus F")
+    plt.ylim(0,2)
+    #plt.xscale("log")
+    #plt.yscale("log")
+    plt.savefig("frb_fluences.pdf")
+    plt.close()
+
+def calc_fluence(F0,Bvals,SNR,w,tsamp,DM,freq,Nants):
```
Comment (profxj) [inline papers/Vvmax/calc_vvmax.py:488]: holy cow, this module is large

maybe break it up if you move code from it into `zdm`

File: papers/Vvmax/calc_vvmax.py [hunk 9]
```diff
@@ -0,0 +1,764 @@
...
+    BeamThetaDeg = data[:,8]
+    oldFnu = data[:,9]
+    zDM = data[:,10]
+    Nants = data[:,11]
+    newFnu = data[:,12]
+    return names,DM,freqMHz,SNR,SNRth,DMG,tsamp,width,BeamThetaDeg,newFnu,zDM,Nants
+
+
+
+global macquartz
```
Comment (profxj) [inline papers/Vvmax/calc_vvmax.py:684]: a `global` here, a `global` there, makes code hard to follow and debug..

File: papers/Vvmax/make_paper_cdf_plot.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
+"""
```
Comment (profxj) [inline papers/Vvmax/make_paper_cdf_plot.py:1]: I haven't reviewed the plots.

**Review by profxj — CHANGES_REQUESTED** [review]
please consider how much of the code could be exposed in the main folder of the
repo, e.g .in `zdm/vvmax`

otherwise, looks great!

File: papers/Vvmax/Data/all_unloc_frb_data.dat (hunk 1 above)
Comment (cwjames1983) [reply papers/Vvmax/Data/all_unloc_frb_data.dat:1]: OK, I'll have a think about that. Ideally, one might open a survey file and do V/Vmax for that survey. I don't think I'll implement that however. But certainly some routines would be useful. Good point, I'll check it out, and maybe re-factorise some routines.

Comment (cwjames1983) [reply papers/Vvmax/Data/all_unloc_frb_data.dat:1]: I've shifted some routines to zdm for vvmax - notably, the actual calculations of V and Vmax! - but the data is in a pretty unique format. I think I'll keeo it that way

File: papers/Vvmax/RLFs.py (hunk 2 above)
Comment (cwjames1983) [reply papers/Vvmax/RLFs.py:1]: I don't think these ones have much value. Mostly, RLFs.py
- keeps constants from the literature, but only for three papers, and only a few constants
- has some specialised routines that integrate the functions over histogram bins
Obviously, they could be useful, but they're also of rather marginal use.

File: papers/Vvmax/Data/all_unloc_frb_data.dat (hunk 1 above)
Comment (cwjames1983) [reply papers/Vvmax/Data/all_unloc_frb_data.dat:1]: So see new file vvmax.py in the zdm directory

File: papers/Vvmax/RLFs.py (hunk 4 above)
Comment (cwjames1983) [reply papers/Vvmax/RLFs.py:36]: I've added proper doc strings to the routines that have been shifted to zdm/vvmax.py . Or, at least, *better* doc strings!

File: papers/Vvmax/calc_vvmax.py (hunk 5 above)
Comment (cwjames1983) [reply papers/Vvmax/calc_vvmax.py]: Yep, this is the part that's been moved. At least, the core of it has.

File: papers/Vvmax/calc_vvmax.py (hunk 6 above)
Comment (cwjames1983) [reply papers/Vvmax/calc_vvmax.py:53]: Yeah I know. Will tell you privately some of this history of this at some point

File: papers/Vvmax/calc_vvmax.py (hunk 7 above)
Comment (cwjames1983) [reply papers/Vvmax/calc_vvmax.py]: Yep, it's in the main code now, sort-of kind0of properly documented

File: papers/Vvmax/make_paper_cdf_plot.py (hunk 10 above)
Comment (cwjames1983) [reply papers/Vvmax/make_paper_cdf_plot.py:1]: I guess they're in the paper :-)

File: papers/Vvmax/RLFs.py (hunk 3 above)
Comment (cwjames1983) [reply papers/Vvmax/RLFs.py:14]: Done

Comment (cwjames1983) [reply papers/Vvmax/RLFs.py:14]: I have defined many of these common values at the top level, since they are used throughout the file

File: papers/Vvmax/calc_vvmax.py (hunk 8 above)
Comment (cwjames1983) [reply papers/Vvmax/calc_vvmax.py:314]: I think I've appropriately broken up useful functions from this in the main module, under vvmax.py.

File: papers/Vvmax/calc_vvmax.py (hunk 9 above)
Comment (cwjames1983) [reply papers/Vvmax/calc_vvmax.py:480]: No globals in the part of the code in the main repo!

**Review by profxj — APPROVED** [review]
excellente
